probe-rs/src/session.rs

```diff
@@ -372,6 +365,34 @@ impl Session {
         self.cores.iter().map(|t| (t.id(), t.core_type())).collect()
     }
 
+    /// Get access to the session when all cores are halted.
+    ///
+    /// Any previously running cores will be resumed once the closure is executed.
+    pub(crate) fn halted_access<R>(
+        &mut self,
+        f: impl FnOnce(&mut Self) -> Result<R, Error>,
+    ) -> Result<R, Error> {
+        let mut resume_state = vec![];
+        for (core, _) in self.list_cores() {
+            let mut c = self.core(core)?;
+            tracing::info!("Core status: {:?}", c.status()?);
+            if !c.core_halted()? {
+                tracing::info!("Halting core...");
+                resume_state.push(core);
+                c.halt(Duration::from_millis(100))?;
```

Comment:
Not something for now, but can cores be halted synchronously?

Reply:
> can cores be halted synchronously

At least Xtensa can do that, but I think this is a common enough multicore problem that others have solutions, too.